### PR TITLE
aichat: prevent page from being summarized multiple times

### PR DESCRIPTION
**Pull request by nullhook**
Resolves https://github.com/brave/brave-browser/issues/36541

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/ai_chat/core/browser/conversation_driver.cc [hunk 1]
```diff
@@ -467,9 +467,11 @@ void ConversationDriver::MaybeSeedOrClearSuggestions() {
...
     // TODO(petemill): ask content fetcher if it knows whether current page is a
     // video.
-    suggestions_.emplace_back(
-        is_video_ ? l10n_util::GetStringUTF8(IDS_CHAT_UI_SUMMARIZE_VIDEO)
-                  : l10n_util::GetStringUTF8(IDS_CHAT_UI_SUMMARIZE_PAGE));
+    if (!has_summarized_) {
+      suggestions_.emplace_back(
+          is_video_ ? l10n_util::GetStringUTF8(IDS_CHAT_UI_SUMMARIZE_VIDEO)
+                    : l10n_util::GetStringUTF8(IDS_CHAT_UI_SUMMARIZE_PAGE));
+    }
```
Comment (nullhook) [inline components/ai_chat/core/browser/conversation_driver.cc:474]: another possible solution is to infer from the length of the conversation_history vector whether it's an initial conversation and add a summary suggestion, rather than maintaining an internal state like this. cc @petemill

Comment (petemill) [reply components/ai_chat/core/browser/conversation_driver.cc:474]: Initial conversation? Not sure what that means exactly. In any case it seems ok to keep internal state about `has_summarized` if it's a requirement of the UI to only show the _summarize_ button if it hasn't been clicked yet.

Comment (petemill) [reply components/ai_chat/core/browser/conversation_driver.cc:474]: I suppose the alternative (perhaps better?) is to check if any history items have `action_type == mojom::ActionType::SUMMARIZE_PAGE` (or `..._VIDEO`)

**Review by petemill — COMMENTED** [review]
Perhaps it's better to check the history?

File: components/ai_chat/core/browser/conversation_driver.cc (hunk 1 above)
Comment (nullhook) [reply components/ai_chat/core/browser/conversation_driver.cc]: checking history now. updated.

File: components/ai_chat/core/browser/conversation_driver.h [hunk 2]
```diff
@@ -10,6 +10,7 @@
 #include <optional>
 #include <string>
 #include <string_view>
+#include <utility>
```
Comment (petemill) [inline components/ai_chat/core/browser/conversation_driver.h:13]: this addition doesn't seem related?

**Review by petemill — APPROVED** [review]

File: components/ai_chat/core/browser/conversation_driver.h (hunk 2 above)
Comment (nullhook) [reply components/ai_chat/core/browser/conversation_driver.h:13]: this was left out in one of your PRs (conversation api one?) which resulted in the following format issue

> ../components/ai_chat/core/browser/conversation_driver.h:147: (cpplint) Add #include <utility> for move [build/include_what_you_use] [4]

**Comment by stephendonner** [comment]
## Verification `IN-PROGRESS` using

```
Brave | 1.68.65 Chromium: 126.0.6478.26 (Official Build) nightly (x86_64)
-- | --
Revision | 2b1404f75647a5223f109757d4551c7a045532b7
OS | macOS Version 11.7.10 (Build 20G1427)
```

First, easily reproduced the issue using both mentioned methods (context menu, open and close tab), using

```
Brave | 1.67.100 Chromium: 125.0.6422.112 (Official Build) beta (x86_64)
-- | --
Revision | 34fcbb6655521b55468af6c63601a8e3c5d24cd2
OS | macOS Version 11.7.10 (Build 20G1427)
```


Screen Shot 2024-05-31 at 3 19 36 PM
